Add problem 'hexhex'

roehrt opened the pull request:
The problem still needs its final polish. Nonetheless, there are two things I couldn't manage:
- [x] Set a memory limit of 1 MB (impossible to do programatically)
- [x] Merge `notes.ipe` and `notes2.ipe`

Before merging the PR we still need to
- [ ] tweak parameters to find a good balance between the jury solution runtimes
- [ ] finally decide on whether or not support python or not.

The memory limit should rule out all BFS solutions and the time limit should rule out the naïve DFS approach :)

roehrt committed:
Add problem 'hexhex'

roehrt committed:
Small textual improvements

EightSQ commented:
AFAIK you cannot set a memory limit directly from the problem. Maybe you can write something like `memory_limit=1024`or `memlimit=1024` in the `domjudge-problem.ini` but [that would be undocumented](https://www.domjudge.org/docs/manual/8.1/problem-format.html). A problem-specific memory limit can be manually set in the jury-side judge interface after the problem is uploaded.

However, I regard enforcing such a tight memory limit as problematic.
First of all, you are essentially forcing Java, which we usually supported this semester, out of competition, as the VM alone requires several hundred megabytes of memory.
Secondly, we always like to see problems being solvable with Python. It being garbage-collected, however, could again be a problem here. I do not have much experience with the memory footprint of Python in CompProg, maybe @PFischbeck knows more.
That being said, you are essentially forcing every solution to be written in C++. I would be honest and write that straight up into the problem description as how different languages work implementation and memory-wise is not really the focus of this lecture/seminar.

I will follow up shortly with feedback about the less technical part.

EightSQ commented on cp22/hpi-fun2/hexhex/problem.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "description": "",
```
A description is mandatory.

EightSQ replied:
Ideally you also mention here that the uploading jury member should set the memory limit of 10 MiB afterwards.

roehrt replied:
So the description is only visible for the jury?

EightSQ commented on cp22/hpi-fun2/hexhex/problem.tex:
```diff
@@ -0,0 +1,43 @@
+
+The first line contains a decimal integer $t$ ($1\leq t\leq 50$) the number of test cases.
+
+Each test case consists of one line containing two hexadecimal integers\footnote{You might want to use \url{https://en.cppreference.com/w/cpp/io/manip/hex} to read them.}
```
C++ is not the only language we use here! (but see my first comment to the PR)

EightSQ commented on cp22/hpi-fun2/hexhex/problem.tex:
```diff
@@ -0,0 +1,43 @@
+Each test case consists of one line containing two hexadecimal integers\footnote{You might want to use \url{https://en.cppreference.com/w/cpp/io/manip/hex} to read them.}
+$h$ and $w$ ($-10^9 \leq h, w \leq 10^9$), the hex-address of your home and the hex-address of the witches, respectively, separated by a space.
+
+It is guaranteed that the sum of $s$ over all test cases does not exceed $200$.
```
(I can imagine what it is but) what is `s`?

roehrt committed:
Fix generator hex/dec

EightSQ commented on cp22/hpi-fun2/hexhex/executables/generator.cpp:
```diff
@@ -0,0 +1,179 @@
+            int max_len = rnd.next(1, min(MAX_LEVEL + 1, MAX_SUM - sum));
+            int w = random_walk(h, max_len);
+            int solution = bfs(h, w);
+            if (solution == -1)
```
Shouldn't this never happen?

roehrt replied:
`rnd.next` has inclusive bounds and if it luckily chooses `MAX_LEVEL+1` the solution needs more than 12 steps (the bfs also stops there).

EightSQ commented on cp22/hpi-fun2/hexhex/executables/generator.cpp:
```diff
@@ -0,0 +1,179 @@
+
+    sample(1, SAMPLE1);
+
+    for (int i = 1; i <= 20; i++)
```
You generate quite a lot of testcases, I think half the amount is sufficient here.

roehrt committed:
Use dec instead of hex

EightSQ commented on cp22/hpi-fun2/hexhex/executables/generator.cpp:
```diff
@@ -0,0 +1,179 @@
+const int MIN_X = -1e9;
+const int MAX_X = 1e9;
+const int MAX_TC = 50;
+const int MAX_SUM = 200;
```
16 testcases with depth 12: 16 * 4^12 = 268435456 steps.
Might be a bit harsh for the time limit of 1 second.

roehrt replied:
I've increased the time limit to be 2.0 seconds and decreased `MAX_SUM` to `150`.

EightSQ commented on cp22/hpi-fun2/hexhex/executables/generator.cpp:
```diff
@@ -0,0 +1,179 @@
+                ++solution;
+                sum += solution;
+                if (sum > MAX_SUM)
+                    return;
```
Please replace this early return logic with an assertion, as you would print less test cases than (MAX_LEVEL * H.size()).

EightSQ reviewed (changes requested):
Hey, all in all I like the problem.
Please 
- [x] look at the single comments I made throughout the PR.

Maybe we'll also have to mention that Memory Limit Exceeded Errors are (occasionally) displayed as "Run Error" on our judge. To prove that a naive DFS is indeed too slow, please
- [x] create a second solution named `solution-dfs-naive.tle.cpp`,

so it can be tested automatically. I'm afraid because we can only change the memory limit manually after the problem is uploaded, we do the same for a BFS solution.

roehrt committed:
Fix a lot

roehrt committed:
Add python solution

roehrt committed:
Add semi-naive solution with pruning

EightSQ commented:
> AFAIK you cannot set a memory limit directly from the problem. Maybe you can write something like `memory_limit=1024`or `memlimit=1024` in the `domjudge-problem.ini` but [that would be undocumented](https://www.domjudge.org/docs/manual/8.1/problem-format.html). A problem-specific memory limit can be manually set in the jury-side judge interface after the problem is uploaded.

Just tried setting both `memory_limit` and `memlimit`. Both are not working. So we have to change it manually after uploading it to the judge.

roehrt committed:
Apply suggestions

roehrt committed:
Increase memory limit

roehrt commented:
I've also added a python solution that is fast enough, if it is executed with pypy. Though pypy uses more memory, it stays under 10 MB as far as I've looked into. According to my calculations a bfs with under 10 MB memory usage is still technically impossible.

roehrt committed:
Half the max sum

roehrt committed:
Rename

roehrt committed:
Let python work and semi-naive dfs fail

EightSQ commented:
@roehrt Right now `make check-all` shows that `solution.py`gives wrong answers.
Locally I have CPython at hand (not pypy), but that should only affect the speed I think.

<details>

```
Checking solution.py
Mismatch on impossible9.in
Wrong answer on line 1 of output (corresponding to line 1 in answer file)
String tokens mismatch
Judge: "-1"
Team: "10"
Mismatch on random5.in
Wrong answer on line 3 of output (corresponding to line 3 in answer file)
String tokens mismatch
Judge: "2"
Team: "4"
Mismatch on random4.in
Wrong answer on line 2 of output (corresponding to line 2 in answer file)
String tokens mismatch
Judge: "4"
Team: "9"
Mismatch on random1.in
Wrong answer on line 11 of output (corresponding to line 11 in answer file)
String tokens mismatch
Judge: "11"
Team: "8"
Mismatch on sample1.in
Wrong answer on line 2 of output (corresponding to line 2 in answer file)
String tokens mismatch
Judge: "3"
Team: "2"
Mismatch on random8.in
Wrong answer on line 10 of output (corresponding to line 10 in answer file)
String tokens mismatch
Judge: "12"
Team: "-1"
Mismatch on random9.in
Wrong answer on line 7 of output (corresponding to line 7 in answer file)
String tokens mismatch
Judge: "12"
Team: "-1"
Mismatch on random7.in
Wrong answer on line 1 of output (corresponding to line 1 in answer file)
String tokens mismatch
Judge: "7"
Team: "-1"
Mismatch on random6.in
Wrong answer on line 1 of output (corresponding to line 1 in answer file)
String tokens mismatch
Judge: "11"
Team: "-1"
Mismatch on random10.in
Wrong answer on line 6 of output (corresponding to line 6 in answer file)
String tokens mismatch
Judge: "10"
Team: "-1"
Mismatch on random3.in
Wrong answer on line 2 of output (corresponding to line 2 in answer file)
String tokens mismatch
Judge: "5"
Team: "8"
Mismatch on random2.in
Wrong answer on line 3 of output (corresponding to line 3 in answer file)
String tokens mismatch
Judge: "11"
Team: "-1"
```
</details>

roehrt committed:
Small corrections

roehrt commented:
Thanks for catching, I didn't know `check-all`! Handy. So the problem was that integer division isn't the same thing for negative numbers in Python and C++. Next time I'll test it properly. I'll make the range positive only.

roehrt committed:
Final parameter tweak

roehrt committed:
Mention RUN-ERROR

roehrt committed:
Use hexspeak samples

roehrt commented:
I've cleaned up the code a bit, adjusted the limits.
Now the python solution stays easily below the time limit (assuming it is executed the same way as in the GCPC https://gcpc.nwerc.eu/system, otherwise with my native system python it's around 40x slower).

roehrt committed:
Polish up solution slides

roehrt committed:
Add problem description

EightSQ commented:
Cool. I only noticed two small things left:
- there is no `-1` test case in the sample (which is okay but makes it harder to debug)
- one animation in the notes is missing (on page 1 both views have all layers enabled).

Oh, and one more thing: the python solution is not working on the judge.
```
Traceback (most recent call last):
  File "solution.py", line 4, in <module>
    from regex import X
ModuleNotFoundError: No module named 'regex'
```
As I do not see the solution using that, please remove it.

And when I look at the python solution I also see that you optimized the input.
```python3
input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline
```
This should not be required when deciding whether or not the problem is solvable in python or not.
FYI the judge executes everything with pypy3 but I do not know which version exactly.

roehrt committed:
Add -1 sample

roehrt committed:
Fix animation

EightSQ commented on cp22/hpi-fun2/hexhex/executables/solution.py:
```diff
@@ -0,0 +1,50 @@
+
+from regex import X
+
+input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline
```
This should not be necessary to solve the problem in Python. Please remove. (I checked, it is still fast enough on the judge).

roehrt replied:
Yep, IO optimization doesn't matter since the input is really small (at most 101 integers). Removed.

EightSQ commented on cp22/hpi-fun2/hexhex/executables/solution.py:
```diff
@@ -0,0 +1,50 @@
+import io
+import os
+
+from regex import X
```
Remove this import, it results in a run error as the `regex` package is not available on the judge.

roehrt committed:
Remove unused import

roehrt committed:
Remove IO optimization

roehrt commented:
Fixed. Thanks for the patient and sharp review of the PR - I appreciate it ^^

EightSQ commented:
Sorry I keep finding small things: `solution-bfs.wa.cpp` I assume should run into out-of-memory, right? Then do not call it `.wa.cpp` as this makes the Makefile wonder that no test case fails when running `make check-all`. I suggest `solution-bfs.rte.cpp` (which has no special meaning at the moment).

It does run, as it should, into OOM on the judge when setting the memory limit. So that's good!

roehrt committed:
Rename to match error source

EightSQ commented:
Alright, some issues came up when testing on the judge:

Both TLE solutions are too fast.
dfs-native 850ms on the anti-pruning testcases (the slowest)
dfs-pruning 740ms on the anti-pruning testcases

The Python solutions runs into OOM and needs up to almost 30 MiB of memory.

roehrt commented:
Unfortunate, I'll increase the memory limit as well as the maximum sum. On my machine they both needed around 5 s..
Let's see if the bfs solution still runs into OOM.

EightSQ commented:
> Unfortunate, I'll increase the memory limit as well as the maximum sum. On my machine they both needed around 5 s..

I know that feeling. On my machine the main solution is half as fast as well. That's why testing on the judge is so important :/
If increasing the memory limit for Python poses too big a risk that BFSs could potentially get through, so be it, contestants this year use Python very rarely anyway.

Increasing the memory limit to 30 MiB still does not let the BFS solution pass. 👍

roehrt commented:
I think my bfs solution was already fairly memory efficient. I tested it: Worst Case scenario would be >100 MB of required memory (I took the RSS value). So I'll increase the limit to 50 MB.

EightSQ commented:
> I think my bfs solution was already fairly memory efficient. I tested it: Worst Case scenario would be > 100 MB of required .emory (I took the RSS value). So I'll increase the limit to 50 MB.

Okay, I'll test it on the judge once again once you increase the bounds and then this is good to go.

roehrt committed:
Parameter tweaks

roehrt commented:
The tle solutions should take twice as long now. I've also added some anti-pruning test cases for different operation orders.

Edit: Oh no! It seems like I managed to create a BFS version using only ~20 MiB. Weirdly, my pypy uses around ~6 MiB. So nonetheless I drop the Python support :( and rollback to 5 MiB. Sorry for the late rollback.

roehrt committed:
Drop python support

EightSQ commented:
Hmm, your `solution-bfs` gives wrong answer:

```
make check-solution-bfs.rte.cpp                                                                   130 ↵
Checking solution-bfs.rte.cpp
Mismatch on random9.in
Wrong answer on line 25 of output (corresponding to line 25 in answer file)
String tokens mismatch
Judge: "11"
Team: "12"
Mismatch on random3.in
Wrong answer on line 3 of output (corresponding to line 3 in answer file)
String tokens mismatch
Judge: "11"
Team: "12"
make: *** [Makefile:109: check-solution-bfs.rte.cpp] Error 1
```

I see you changed a `long long` to `int`. Maybe this now runs into overflows?

EightSQ commented:
Much worse: now every solution is too slow.
`solution.cpp` uses 1006ms on the judge. We aim to stay comfortably within the time limit, so increasing both TC and SUM might have been too generous. (TC=200 means 2.6*10^8 steps after all, which is too much).

What confuses me is that now your TLE solutions and the "correct" solutions are barely distinguishable runtime-wise ...

EightSQ commented:
Your new anti-pruning test cases seem to be wayyyy to slow in some cases:
```
Timing solution.cpp
anti-pruning1.in:             1.524s
anti-pruning10.in:            0.021s
anti-pruning11.in:            0.024s
anti-pruning12.in:            0.018s
anti-pruning2.in:             1.694s
anti-pruning3.in:             1.485s
anti-pruning4.in:             0.010s
anti-pruning5.in:             0.011s
anti-pruning6.in:             0.014s
anti-pruning7.in:             0.014s
anti-pruning8.in:             0.015s
anti-pruning9.in:             0.015s
```

EightSQ commented:
In case you cannot continue working on this until the contest start in a few hours, I will be using commit 1858a427a7adb046572e57183e3f80e7f6d721e2 with the tweak of setting MAX_TC=150 and MAX_SUM=300. This makes Python too slow in one random test case (see attached image) and both TLE solutions need 1.4s on an anti-pruning test case. As a result, I will be setting the memory limit to 5 MiB.

It might be possible to still pass with custom-made hex fast-io but I doubt anyone will be eager enough to build that (or find copypasta from the internet) instead of having the right idea for the solution.

The correct cpp solution gets through comfortably (175ms).

![Screenshot 2022-08-05 at 00 53 57](https://user-images.githubusercontent.com/1622854/182968712-58743afa-5204-4b9c-9cde-3f596bb01468.png)

roehrt committed:
Fix bfs

roehrt commented:
> Hmm, your `solution-bfs` gives wrong answer:
> 
> ```
> make check-solution-bfs.rte.cpp                                                                   130 ↵
> Checking solution-bfs.rte.cpp
> Mismatch on random9.in
> Wrong answer on line 25 of output (corresponding to line 25 in answer file)
> String tokens mismatch
> Judge: "11"
> Team: "12"
> Mismatch on random3.in
> Wrong answer on line 3 of output (corresponding to line 3 in answer file)
> String tokens mismatch
> Judge: "11"
> Team: "12"
> make: *** [Makefile:109: check-solution-bfs.rte.cpp] Error 1
> ```
> 
> I see you changed a `long long` to `int`. Maybe this now runs into overflows?

Yeah, my memory saving optimization was wrong.

> Your new anti-pruning test cases seem to be wayyyy to slow in some cases:

Depending on your dfs traversal order only 3 of them should provoke the tle.

> What confuses me is that now your TLE solutions and the "correct" solutions are barely distinguishable runtime-wise ...

Weird! My local benchmarks:

<details>
<pre>
❯  for i in (seq 10); /usr/bin/time -f "max mem=%M elapsed=%E cpu.sys=%S user=%U" ./executables/solution > my$(echo $i).ans  < build/testcases/random$(echo $i).in; end
max mem=3500 elapsed=0:00.34 cpu.sys=0.00 user=0.32
max mem=3492 elapsed=0:00.34 cpu.sys=0.00 user=0.34
max mem=3448 elapsed=0:00.17 cpu.sys=0.00 user=0.17
max mem=3500 elapsed=0:00.35 cpu.sys=0.00 user=0.35
max mem=3500 elapsed=0:00.39 cpu.sys=0.00 user=0.39
max mem=3444 elapsed=0:00.41 cpu.sys=0.00 user=0.41
max mem=3500 elapsed=0:00.13 cpu.sys=0.00 user=0.13
max mem=3500 elapsed=0:00.20 cpu.sys=0.00 user=0.20
max mem=3420 elapsed=0:00.30 cpu.sys=0.00 user=0.30
max mem=3488 elapsed=0:00.34 cpu.sys=0.00 user=0.34

❯  for i in (seq 10); /usr/bin/time -f "max mem=%M elapsed=%E cpu.sys=%S user=%U" ./executables/solution-dfs-naive.tle > my$(echo $i).ans  < build/testcases/random$(echo $i).in; end
max mem=3508 elapsed=0:01.58 cpu.sys=0.00 user=1.58
max mem=3488 elapsed=0:01.41 cpu.sys=0.00 user=1.41
max mem=3504 elapsed=0:01.40 cpu.sys=0.00 user=1.40
max mem=3424 elapsed=0:01.61 cpu.sys=0.00 user=1.61
max mem=3488 elapsed=0:01.33 cpu.sys=0.00 user=1.33
max mem=3500 elapsed=0:01.33 cpu.sys=0.00 user=1.32
max mem=3416 elapsed=0:01.90 cpu.sys=0.00 user=1.90
max mem=3444 elapsed=0:01.28 cpu.sys=0.00 user=1.28
max mem=3492 elapsed=0:01.38 cpu.sys=0.00 user=1.38
max mem=3496 elapsed=0:01.82 cpu.sys=0.00 user=1.82

❯  for i in (seq 10); /usr/bin/time -f "max mem=%M elapsed=%E cpu.sys=%S user=%U" ./executables/solution-dfs-pruning.tle > my$(echo $i).ans  < build/testcases/anti-pruning$(echo $i).in; end
max mem=3420 elapsed=0:03.30 cpu.sys=0.00 user=3.27
max mem=3420 elapsed=0:03.60 cpu.sys=0.00 user=3.60
max mem=3420 elapsed=0:03.88 cpu.sys=0.00 user=3.88
max mem=3504 elapsed=0:04.04 cpu.sys=0.00 user=4.04
max mem=3532 elapsed=0:03.27 cpu.sys=0.00 user=3.27
max mem=3532 elapsed=0:03.11 cpu.sys=0.00 user=3.11
max mem=3492 elapsed=0:00.00 cpu.sys=0.00 user=0.00
max mem=3488 elapsed=0:00.00 cpu.sys=0.00 user=0.00
max mem=3504 elapsed=0:00.00 cpu.sys=0.00 user=0.00
max mem=3444 elapsed=0:00.01 cpu.sys=0.00 user=0.00
</pre>
</details>

roehrt committed:
Fix test case generator

roehrt commented:
> What confuses me is that now your TLE solutions and the "correct" solutions are barely distinguishable runtime-wise ...

I accidentally broke the tc generator! (integer overflow, when I generalized the operation in the anti-pruning test cases) I fixed it. Now the intended solution shouldn't hit the time limit and pass easily.
Even though everything should work now, it seems a bit late now..

roehrt committed:
Add multiplication test cases again

EightSQ commented:
> 

Not a problem, last minute changes are what we are paid for. :)
Everything seems to work fine, even though the BFS needs slightly more than 50 MiB I set a memory limit of 5 MiB as you requested.

EightSQ reviewed (approved):
Looks good to me now
